client/widmgr/widmgr.go

```diff
@@ -4,80 +4,367 @@
+				AllocID:      m.allocID,
+				TaskName:     taskName,
+				IdentityName: widspec.Name,
+			})
+		}
+	}
+
+	renewNow := false
+	minExp := time.Now().Add(30 * time.Hour) // set high default expiration
+
+	for taskName, wids := range m.widSpecs {
+		for _, wid := range wids {
+			if wid.TTL == 0 {
+				// No ttl, so no need to renew it
+				continue
+			}
+
+			//FIXME make this less ugly
+			token := m.get(cstructs.TaskIdentity{
+				TaskName:     taskName,
+				IdentityName: wid.Name,
+			})
+			if token == nil {
+				// Missing a signature, treat this case as already expired so we get a
+				// token ASAP
+				m.logger.Debug("missing token for identity", "identity", wid.Name)
+				renewNow = true
+				continue
+			}
+
+			if token.Expiration.Before(minExp) {
+				minExp = token.Expiration
+			}
+		}
+	}
+
+	if len(reqs) == 0 {
+		m.logger.Trace("no workload identities expire")
+		return
+	}
+
+	var wait time.Duration
+	if !renewNow {
+		wait = helper.ExpiryToRenewTime(minExp, time.Now, m.minWait)
+	}
+
+	timer, timerStop := helper.NewStoppedTimer()
+	defer timerStop()
+
+	var retry uint64
+
+	for {
+		// we need to handle stopCtx.Err() and manually stop the subscribers
+		if err := m.stopCtx.Err(); err != nil {
+			//TODO Close watchers
+			return
+		}
+
+		m.logger.Debug("waiting to renew identities", "num", len(reqs), "wait", wait)
+		timer.Reset(wait)
+		select {
+		case <-timer.C:
+			m.logger.Trace("getting new signed identities", "num", len(reqs))
+		case <-m.stopCtx.Done():
+			//TODO Close watchers
+			return
+		}
+
+		// Renew all tokens together since its cheap
+		tokens, err := m.signer.SignIdentities(m.minIndex, reqs)
+		if err != nil {
+			retry++
+			wait = helper.Backoff(m.minWait, time.Hour, retry) + helper.RandomStagger(m.minWait)
+			m.logger.Error("error renewing workload identities", "error", err, "next", wait)
+			continue
+		}
+
+		if len(tokens) == 0 {
+			retry++
+			wait = helper.Backoff(m.minWait, time.Hour, retry) + helper.RandomStagger(m.minWait)
+			m.logger.Error("error renewing workload identities", "error", "no tokens", "next", wait)
+			continue
+		}
+
+		// Reset next expiration time
+		minExp = time.Time{}
+
+		for _, token := range tokens {
+			id := cstructs.TaskIdentity{
+				TaskName:     token.TaskName,
+				IdentityName: token.IdentityName,
+			}
+
+			// Set for getters
+			m.lastTokenLock.Lock()
+			m.lastToken[id] = token
+			m.lastTokenLock.Unlock()
+
+			// Send to watchers
+			m.send(id, token)
+
+			// Set next expiration time
+			if minExp.IsZero() {
+				minExp = token.Expiration
+			} else if token.Expiration.Before(minExp) {
+				minExp = token.Expiration
+			}
+		}
+
+		// Success! Set next renewal and reset retries
+		wait = helper.ExpiryToRenewTime(minExp, time.Now, m.minWait)
+		retry = 0
+	}
+}
+
+func (m *WIDMgr) send(id cstructs.TaskIdentity, token *structs.SignedWorkloadIdentity) {
+	m.watchersLock.Lock()
+	defer m.watchersLock.Unlock()
```

Comment:
`send` has only one caller, so it'd be more efficient to lock in the caller and then note that this method must be called while holding the lock (elsewhere with this pattern we've even changed the method name to something like `sendLocked`... unfortunately the type system doesn't give us a lot of help here :crying_cat_face: )

Reply:
good call, I modified in [e8ef776](https://github.com/hashicorp/nomad/pull/18431/commits/e8ef7763782f937b3915c163134604544919f735)